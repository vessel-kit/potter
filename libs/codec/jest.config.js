const base = require('../../common/jest.config');
const packageJson = require('./package.json')

module.exports = {
  ...base,
  displayName: packageJson.name
};
