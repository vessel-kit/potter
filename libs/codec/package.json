{
  "name": "@potter/codec",
  "private": true,
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "version": "1.0.0-a.0",
  "scripts": {
    "build": "tsc -p .",
    "dev": "tsc -p . --watch"
  },
  "dependencies": {
    "cids": "^0.8.0",
    "decko": "^1.2.0",
    "fp-ts": "^2.0.0",
    "io-ts": "^2.2.7",
    "multibase": "^1.0.1"
  },
  "devDependencies": {
    "@types/multibase": "^0.6.0"
  }
}
