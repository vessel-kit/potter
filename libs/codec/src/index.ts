export * from './enum-of';
export * from './split-string';
export * from './decode-throw';
export * from './ceramic-document-id';
export * from './decode-pipe';
export * from './normalize-record';
export * from './record-wrap';
export * from './simple-codec'

export * from './codecs/date.number.codec';
export * from './codecs/cid.ipld.codec';
export * from './codecs/cid.string.codec';
export * from './codecs/cid.object.codec';
export * from './codecs/ceramic-document-id.cid.codec'
export * from './codecs/ceramic-document-id.string.codec';
export * from './codecs/buffer.multibase.codec';
export * from './codecs/date.iso8601.codec';
export * from './codecs/fast-patch-operation.json.codec'
