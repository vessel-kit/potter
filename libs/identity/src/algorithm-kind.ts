import { InvalidAlgorithmKindError } from './invalid-algorithm-kind.error';

export enum AlgorithmKind {
  ES256K = 'ES256K',
  EdDSA = 'EdDSA',
}

/* istanbul ignore next */
export namespace AlgorithmKind {
  export function fromString(input: string) {
    const searching = input as AlgorithmKind;
    switch (searching) {
      case AlgorithmKind.EdDSA:
        return AlgorithmKind.EdDSA;
      case AlgorithmKind.ES256K:
        return AlgorithmKind.ES256K;
      default:
        throw new InvalidAlgorithmKindError(searching);
    }
  }
}
