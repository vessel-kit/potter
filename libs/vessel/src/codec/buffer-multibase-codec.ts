import * as t from 'io-ts';
import multibase from 'multibase';

export const BufferMultibaseCodec = new t.Type<Buffer, string, string>(
  'BufferMultibaseCodec',
  (input: unknown): input is Buffer => Buffer.isBuffer(input) && Boolean(multibase.isEncoded(input as Buffer)),
  input => t.success(multibase.decode(input)),
  (a: Buffer) => multibase.encode('base58btc', a).toString(),
);
