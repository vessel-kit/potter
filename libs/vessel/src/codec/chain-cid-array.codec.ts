import * as t from 'io-ts';
import { Chain } from '../chain';
import CID from 'cids';

export const ChainCidArrayCodec = new t.Type<Chain, CID[], CID[]>(
  'Chain-CIDArray',
  (input: unknown): input is Chain => input instanceof Chain,
  input => {
    return t.success(new Chain(input));
  },
  (a: Chain) => a.log,
);
