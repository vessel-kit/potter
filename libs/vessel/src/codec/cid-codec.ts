import * as t from 'io-ts';
import CID from 'cids';

export const CidCodec = new t.Type<CID, string, string>(
  'CID',
  (input: unknown): input is CID => CID.isCID(input),
  (input, context) => {
    try {
      return t.success(new CID(input));
    } catch (e) {
      t.failure(input, context, 'Invalid CID');
    }
  },
  (a: CID) => CID.toString(),
);
