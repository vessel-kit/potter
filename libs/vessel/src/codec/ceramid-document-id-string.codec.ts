import * as t from 'io-ts';
import { CeramicDocumentId } from '@potter/codec';

export const CeramicDocumentIdStringCodec = new t.Type<CeramicDocumentId, string, string>(
  'CeramicDocumentIdString',
  (input: unknown): input is CeramicDocumentId => input instanceof CeramicDocumentId,
  (input, context) => {
    try {
      return t.success(CeramicDocumentId.fromString(input));
    } catch (e) {
      t.failure(input, context, 'Invalid Ceramic document id');
    }
  },
  (a: CeramicDocumentId) => a.toString(),
);
