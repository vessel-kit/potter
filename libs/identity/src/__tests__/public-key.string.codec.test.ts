import '@relmify/jest-fp-ts';
import { AlgorithmKind } from '../algorithm-kind';
import { PrivateKeyFactory } from '../private-key.factory';
import { PublicKeyStringCodec } from '../public-key.string.codec';
import { BytesMultibaseCodec } from "@vessel-kit/codec";

const privateKeyFactory = new PrivateKeyFactory();
const material = new Uint8Array([1, 2, 3]);

describe('is', () => {
  test('ok', async () => {
    const privateKey = privateKeyFactory.fromSeed(AlgorithmKind.ES256K, material);
    const publicKey = await privateKey.publicKey();
    expect(PublicKeyStringCodec.is(publicKey)).toBeTruthy();
  });

  test('not ok', async () => {
    expect(PublicKeyStringCodec.is({ garbage: true })).toBeFalsy();
    expect(PublicKeyStringCodec.is('just stringÂ£')).toBeFalsy();
  });
});

describe('encode', () => {
  test('ES256K', async () => {
    const privateKey = privateKeyFactory.fromSeed(AlgorithmKind.ES256K, material);
    const publicKey = await privateKey.publicKey();
    expect(PublicKeyStringCodec.encode(publicKey)).toMatchSnapshot();
  });
  test('Ed25519', async () => {
    const privateKey = privateKeyFactory.fromSeed(AlgorithmKind.EdDSA, material);
    const publicKey = await privateKey.publicKey();
    expect(PublicKeyStringCodec.encode(publicKey)).toMatchSnapshot();
  });
});

describe('decode', () => {
  test('ES256K', async () => {
    const privateKey = privateKeyFactory.fromSeed(AlgorithmKind.ES256K, material);
    const publicKey = await privateKey.publicKey();
    expect(PublicKeyStringCodec.decode(PublicKeyStringCodec.encode(publicKey))).toEqualRight(publicKey);
  });
  test('Ed25519', async () => {
    const privateKey = privateKeyFactory.fromSeed(AlgorithmKind.EdDSA, material);
    const publicKey = await privateKey.publicKey();
    expect(PublicKeyStringCodec.decode(PublicKeyStringCodec.encode(publicKey))).toEqualRight(publicKey);
  });
  test('garbage', () => {
    expect(PublicKeyStringCodec.decode('garbage')).toBeLeft();
    const garbage = BytesMultibaseCodec('base16').encode(material)
    expect(PublicKeyStringCodec.decode(garbage)).toBeLeft();
  });
});
