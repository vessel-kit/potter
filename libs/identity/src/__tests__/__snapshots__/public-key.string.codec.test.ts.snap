// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encode ES256K 1`] = `"zQ3shkAX9qxg2DUj3gd9tv7fmbJAuZSnr6W5ELTJJc1uUfNsf"`;

exports[`encode Ed25519 1`] = `"z6LSqVDApMd14pj9gYfNFR6hSBnziMgxePPq8mzoWQHkFfBM"`;

exports[`encode EdDSA 1`] = `"z6LSqVDApMd14pj9gYfNFR6hSBnziMgxePPq8mzoWQHkFfBM"`;
